In [1]:
# All Libraries required for this lab are listed below. The libraries pre-installed on Skills Network Labs are commented. 
# If you run this notebook on a different environment, e.g. your desktop, you may need to uncomment and install certain libraries.

#!pip install numpy==1.21.4
#!pip install pandas==1.3.4
#!pip install keras==2.1.6

We will first import the necessary libraries.

In [2]:
import pandas as pd
import numpy as np

Let's import our data now.

In [3]:
concrete_data = pd.read_csv('https://cocl.us/concrete_data')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [4]:
concrete_data.shape

(1030, 9)

In [5]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

We have 1030 tupples and there are no nulls in our data. Great. 
Let's separate our dataframe into two: one for our predictors and another one for our target.

In [6]:
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

In [7]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [8]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

In [9]:
n_cols = predictors.shape[1] # number of predictors

So far everything is ok. Let's move along.

## Use the Keras library to build a neural network 

Use the Keras library to build a neural network with the following:
- One hidden layer of 10 nodes, and a ReLU activation function
- Use the adam optimizer and the mean squared error  as the loss function.
1. Randomly split the data into a training and test sets by holding 30% of the data for testing. You can use the train_test_split helper function from Scikit-learn.
2. Train the model on the training data using 50 epochs.
3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. You can use the mean_squared_error function from Scikit-learn.
4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.
5. Report the mean and the standard deviation of the mean squared errors.


## Import keras and sklearn libraries

In [10]:
import keras
from keras.models import Sequential
from keras.layers import Dense

Using TensorFlow backend.
/home/jupyterlab/conda/envs/python/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:516: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/home/jupyterlab/conda/envs/python/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:517: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/home/jupyterlab/conda/envs/python/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:518: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/home/jupyterlab/con

We will import additional libraries as well.

In [11]:
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split

/home/jupyterlab/conda/envs/python/lib/python3.7/site-packages/sklearn/utils/validation.py:37: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
  LARGE_SPARSE_SUPPORTED = LooseVersion(scipy_version) >= '0.14.0'


## **Part A.**

Let's build and compile our model.

In [12]:
model_A = Sequential()
model_A.add(Dense(10, activation='relu', input_shape=(n_cols,))) # 1 hidden layer of 10 nodes
model_A.add(Dense(1))
model_A.compile(optimizer='adam', loss='mean_squared_error') # adam optimizer and the mean squared error as the loss function

We will randomly split the data into a training and test sets by holding 30% of the data for testing. 

In [13]:
X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=42)

Fit the model. We will train the model on the training data using 50 epochs.

In [14]:
model_A.fit(X_train, y_train, epochs=50, verbose=0)

2023-03-22 05:54:40.328271: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 AVX512F FMA
2023-03-22 05:54:40.333437: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2494075000 Hz
2023-03-22 05:54:40.334117: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x55568314c350 executing computations on platform Host. Devices:
2023-03-22 05:54:40.334178: I tensorflow/compiler/xla/service/service.cc:175]   StreamExecutor device (0): <undefined>, <undefined>
2023-03-22 05:54:40.400980: W tensorflow/compiler/jit/mark_for_compilation_pass.cc:1412] (One-time warning): Not using XLA:CPU for cluster because envvar TF_XLA_FLAGS=--tf_xla_cpu_global_jit was not set.  If you want XLA:CPU, either set that envvar, or use experimental_jit_scope to enable XLA:CPU.  To confirm that XLA is active, pass --vmodule=xla_compilation_cache=1 (as a proper command-line fl

Now we will evaluate the model on the test data and compute the mean squared error between the predicted value and the actual value. 

In [15]:
model_A_eval = model_A.evaluate(X_test, y_test)
model_A_eval

309/309 [==============================] - 0s 101us/step


469.2619698039922

In [16]:
y_pred = model_A.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
mse

469.2619161022984

Now we will repeat the previous steps 50 times in order to get 50 mean squared errors.

In [17]:
mean_squared_errors = []
for i in range(0,50):
    # split the data
    X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=i)
    # fit the model with our data
    model_A.fit(X_train, y_train, epochs=50, verbose=0)
    # evaluate the model
    model_eval = model_A.evaluate(X_test, y_test, verbose=0)
    # calculate the mse
    y_pred = model_A.predict(X_test)
    mse = mean_squared_error(y_test, y_pred)
    # append our mse to our list
    mean_squared_errors.append(mse)

Let's check out our list of mean squared errors.

Report the mean and the standard deviation of the mean squared errors.

In [18]:
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)

In [19]:
mean

51.711954125356016

In [20]:
standard_deviation

17.040345571772622

## **Part B.**

First, let's normalize our data.

In [21]:
predictors_normalized = (predictors - predictors.mean()) / predictors.std()
predictors_normalized.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [22]:
# Build and compile the model
model_B = Sequential()
model_B.add(Dense(10, activation='relu', input_shape=(n_cols,))) # 1 hidden layer of 10 nodes
model_B.add(Dense(1))
model_B.compile(optimizer='adam', loss='mean_squared_error') # adam optimizer and the mean squared error as the loss function

In [23]:
# Split the data
X_train, X_test, y_train, y_test = train_test_split(predictors_normalized, target, test_size=0.3, random_state=42)

In [24]:
# Fit the model
model_B.fit(X_train, y_train, epochs=50, verbose=0)

In [25]:
# Evaluate the model
model_B_eval = model_B.evaluate(X_test, y_test)
model_B_eval

309/309 [==============================] - 0s 128us/step


380.5197938591917

In [26]:
# Find the MSE
y_pred = model_B.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
mse

380.5197759648453

In [27]:
# Repeat the process 50 times
mean_squared_errors_2 = []
for i in range(0,50):
    # split the data
    X_train, X_test, y_train, y_test = train_test_split(predictors_normalized, target, test_size=0.3, random_state=i)
    # fit the model with our data
    model_B.fit(X_train, y_train, epochs=50, verbose=0)
    # evaluate the model
    model_eval = model_B.evaluate(X_test, y_test, verbose=0)
    # calculate the mse
    y_pred = model_B.predict(X_test)
    mse = mean_squared_error(y_test, y_pred)
    # append our mse to our list
    mean_squared_errors_2.append(mse)

In [28]:
# Calculating the mean and std
mean2 = np.mean(mean_squared_errors_2)
standard_deviation2 = np.std(mean_squared_errors_2)

In [29]:
mean2

41.94114665658969

In [30]:
standard_deviation2

24.100195964861587

## **Part C.**

In [31]:
# Build and compile the model
model_C = Sequential()
model_C.add(Dense(10, activation='relu', input_shape=(n_cols,))) # 1 hidden layer of 10 nodes
model_C.add(Dense(1))
model_C.compile(optimizer='adam', loss='mean_squared_error') # adam optimizer and the mean squared error as the loss function

In [32]:
# Split the data
X_train, X_test, y_train, y_test = train_test_split(predictors_normalized, target, test_size=0.3, random_state=42)

In [33]:
# Fit the model. Use 100 epochs this time for training
model_C.fit(X_train, y_train, epochs=100, verbose=0)

In [34]:
# Evaluate the model
model_C_eval = model_C.evaluate(X_test, y_test)
model_C_eval

309/309 [==============================] - 0s 170us/step


150.09656625889653

In [35]:
# Find the MSE
y_pred = model_C.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
mse

150.09656575728866

In [36]:
# Repeat the process 50 times
mean_squared_errors_3 = []
for i in range(0,50):
    # split the data
    X_train, X_test, y_train, y_test = train_test_split(predictors_normalized, target, test_size=0.3, random_state=i)
    # fit the model with our data. Use 100 epochs for training
    model_C.fit(X_train, y_train, epochs=100, verbose=0)
    # evaluate the model
    model_eval = model_C.evaluate(X_test, y_test, verbose=0)
    # calculate the mse
    y_pred = model_C.predict(X_test)
    mse = mean_squared_error(y_test, y_pred)
    # append our mse to our list
    mean_squared_errors_3.append(mse)

In [37]:
# Calculating the mean and std
mean3 = np.mean(mean_squared_errors_3)
standard_deviation3 = np.std(mean_squared_errors_3)

In [38]:
mean3

44.877676036864074

In [39]:
standard_deviation3

9.024181040353426

## **Part D.**

In [40]:
# Build and compile the model. This model uses 3 hidden layers as opposed to 1 like in previous models
model_D = Sequential()
model_D.add(Dense(10, activation='relu', input_shape=(n_cols,)))
model_D.add(Dense(10, activation='relu'))
model_D.add(Dense(10, activation='relu'))
model_D.add(Dense(1))
model_D.compile(optimizer='adam', loss='mean_squared_error') # adam optimizer and the mean squared error as the loss function

In [41]:
# Split the data
X_train, X_test, y_train, y_test = train_test_split(predictors_normalized, target, test_size=0.3, random_state=42)

In [42]:
# Fit the model
model_D.fit(X_train, y_train, epochs=50, verbose=0)

In [43]:
# Evaluate the model
model_D_eval = model_D.evaluate(X_test, y_test)
model_D_eval

309/309 [==============================] - 0s 226us/step


109.8684687938505

In [44]:
# Find the MSE
y_pred = model_D.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
mse

109.8684723010953

In [45]:
# Repeat the process 50 times
mean_squared_errors_4 = []
for i in range(0,50):
    # split the data
    X_train, X_test, y_train, y_test = train_test_split(predictors_normalized, target, test_size=0.3, random_state=i)
    # fit the model with our data
    model_D.fit(X_train, y_train, epochs=50, verbose=0)
    # evaluate the model
    model_eval = model_D.evaluate(X_test, y_test,verbose=0)
    # calculate the mse
    y_pred = model_D.predict(X_test)
    mse = mean_squared_error(y_test, y_pred)
    # append our mse to our list
    mean_squared_errors_4.append(mse)

In [46]:
# Calculating the mean and std
mean4 = np.mean(mean_squared_errors_4)
standard_deviation4 = np.std(mean_squared_errors_4)

In [47]:
mean4

25.603135131939844

In [48]:
standard_deviation4

7.089338931915645

## **Conclusion**

In [49]:
print("Mean Summary:")
print("Model A: " + str(mean))
print("Model B: " + str(mean2))
print("Model C: " + str(mean3))
print("Model D: " + str(mean4))
print('\n')
print("Standard Deviation Summary:")
print("Model A: " + str(standard_deviation))
print("Model B: " + str(standard_deviation2))
print("Model C: " + str(standard_deviation3))
print("Model D: " + str(standard_deviation4))

Mean Summary:
Model A: 51.711954125356016
Model B: 41.94114665658969
Model C: 44.877676036864074
Model D: 25.603135131939844


Standard Deviation Summary:
Model A: 17.040345571772622
Model B: 24.100195964861587
Model C: 9.024181040353426
Model D: 7.089338931915645


**B. How does the mean of the mean squared errors compare to that from Step A?**

The mse is smaller than the mse from Step A.

**C. How does the mean of the mean squared errors compare to that from Step B?**

The mse is slighly larger than the mse from Step B.

**D. How does the mean of the mean squared errors compare to that from Step B?**

The mse is much smaller than the mse from Step B.